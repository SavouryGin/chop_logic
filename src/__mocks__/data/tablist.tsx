import React from 'react';
import { TabItem } from 'types';

export const testTabs: TabItem[] = [
  { tabContent: <>Tab content 1</>, tabId: 'tab-1', tabTitle: { en: 'Title 1', ru: 'Заголовок 1' } },
  { tabContent: <>Tab content 2</>, tabId: 'tab-2', tabTitle: { en: 'Title 2', ru: 'Заголовок 2' } },
  { tabContent: <>Tab content 3</>, tabId: 'tab-3', tabTitle: { en: 'Title 3', ru: 'Заголовок 3' } },
  { tabContent: <>Tab content 4</>, tabId: 'tab-4', tabTitle: { en: 'Title 4', ru: 'Заголовок 4' } },
];
