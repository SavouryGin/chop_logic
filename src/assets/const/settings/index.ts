import { SelectEntity } from 'types';

export const languageOptions: SelectEntity[] = [
  { option: { en: 'English', ru: 'Английский' }, value: 'en' },
  { option: { en: 'Russian', ru: 'Русский' }, value: 'ru' },
];

export const externalLinks: { [key in string]: string } = {
  telegram: 'https://telegram.me/savoury_gin',
  gitHub: 'https://github.com/SavouryGin',
  mail: 'mailto:Dmitrii.Suroviagin@gmail.com',
};
