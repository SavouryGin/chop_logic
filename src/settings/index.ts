import { SelectEntity } from 'types';

export const languageOptions: SelectEntity[] = [
  { option: { en: 'English', ru: 'Английский' }, value: 'en' },
  { option: { en: 'Russian', ru: 'Русский' }, value: 'ru' },
];

export const languageStringOptions = ['en', 'ru'];

export const externalLinks: { [key in string]: string } = {
  telegram: 'https://telegram.me/savoury_gin',
  gitHub: 'https://github.com/SavouryGin',
  linkedIn: 'https://www.linkedin.com/in/dmitrii-suroviagin/',
  mail: 'mailto:Dmitrii.Suroviagin@gmail.com',
  facebook: 'https://www.facebook.com/dmitrii.suroviagin',
};
