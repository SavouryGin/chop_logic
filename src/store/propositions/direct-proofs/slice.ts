import converter from 'logic/propositions/converter';
import executor from 'logic/propositions/executor';
import replacer from 'logic/propositions/replacer';
import validator from 'logic/propositions/validator';
import { DirectProofsTableItem, PropositionsDirectProofsFlag, PropositionsDirectProofsInitialState } from './interfaces';
import { Guid } from 'guid-typescript';
import { PayloadAction, createSlice } from '@reduxjs/toolkit';

export const propositionsDPInitialState: PropositionsDirectProofsInitialState = {
  flags: {
    isPremiseOpened: false,
    isImplicationCreationOpened: false,
    isImplicationDistributionOpened: false,
    isContradictionRealizationOpened: false,
    isReplacerFormOpened: false,
    isConfirmDeletePopupOpened: false,
  },
  tableData: [],
  selectedIds: [],
  dependentItems: [],
};

export const propositionsDPSlice = createSlice({
  name: 'propositionsDP',
  initialState: propositionsDPInitialState,
  reducers: {
    setUpFlag: (state, action: PayloadAction<{ flag: PropositionsDirectProofsFlag; value: boolean }>) => {
      const { flag, value } = action.payload;
      state.flags[flag] = value;
    },

    setSelectedIds: (state, action: PayloadAction<string[]>) => {
      state.selectedIds = action.payload;
    },

    setTableData: (state, action: PayloadAction<DirectProofsTableItem[]>) => {
      state.tableData = action.payload;
    },

    setDependentItems: (state, action: PayloadAction<DirectProofsTableItem[]>) => {
      state.dependentItems = action.payload;
    },

    addPremise: (state, action: PayloadAction<string>) => {
      const rawInput = action.payload;
      const expression = converter.convertStringToExpression(rawInput);
      const formula = converter.convertExpressionToFormula(expression);
      const friendlyExpression = converter.convertFormulaToUserFriendlyExpression(formula);
      const step = state.tableData.length + 1;
      const id = Guid.create().toString();
      const newItem: DirectProofsTableItem = {
        id,
        step,
        rawInput,
        expression,
        friendlyExpression,
        formula,
        comment: { en: 'Premise', ru: 'Посылка' },
      };
      state.tableData = [...state.tableData, newItem];
    },

    deleteSteps: (state, _action: PayloadAction<{ isConfirmed: boolean }>) => {
      return state;
    },

    reiterateStep: (state) => {
      const selectedStep = state.tableData.find((item) => item.id === state.selectedIds[0]);
      if (!selectedStep) {
        return state;
      }
      const index = state.tableData.length + 1;
      const newStep = {
        ...selectedStep,
        id: Guid.create().toString(),
        step: index,
        comment: { en: `Reiter. ${selectedStep.step}`, ru: `Повтор ${selectedStep.step}` },
      };

      state.selectedIds = [];
      state.tableData = [...state.tableData, newStep];
    },

    createImplication: (state, action: PayloadAction<{ firstVariable: string; secondVariable: string }>) => {
      const { firstVariable, secondVariable } = action.payload;
      const expression = converter.convertToICExpression(firstVariable, secondVariable);
      const formula = converter.convertExpressionToFormula(expression);
      const friendlyExpression = converter.convertFormulaToUserFriendlyExpression(formula);
      const step = state.tableData.length + 1;
      const id = Guid.create().toString();
      const newItem: DirectProofsTableItem = {
        step,
        id,
        rawInput: `${firstVariable}, ${secondVariable}`,
        expression,
        friendlyExpression,
        formula,
        comment: { en: 'IC', ru: 'ВИ' },
      };
      state.selectedIds = [];
      state.tableData = [...state.tableData, newItem];
    },

    createImplicationDistribution: (
      state,
      action: PayloadAction<{ firstVariable: string; secondVariable: string; thirdVariable: string }>,
    ) => {
      const { firstVariable, secondVariable, thirdVariable } = action.payload;
      const expression = converter.convertToIDExpression(firstVariable, secondVariable, thirdVariable);
      const formula = converter.convertExpressionToFormula(expression);
      const friendlyExpression = converter.convertFormulaToUserFriendlyExpression(formula);
      const step = state.tableData.length + 1;
      const id = Guid.create().toString();
      const newItem: DirectProofsTableItem = {
        step,
        id,
        rawInput: `${firstVariable}, ${secondVariable}, ${thirdVariable}`,
        expression,
        friendlyExpression,
        formula,
        comment: { en: 'ID', ru: 'ДИ' },
      };
      state.selectedIds = [];
      state.tableData = [...state.tableData, newItem];
    },

    createContradictionRealization: (state, action: PayloadAction<{ firstVariable: string; secondVariable: string }>) => {
      const { firstVariable, secondVariable } = action.payload;
      const expression = converter.convertToCRExpression(firstVariable, secondVariable);
      const formula = converter.convertExpressionToFormula(expression);
      const friendlyExpression = converter.convertFormulaToUserFriendlyExpression(formula);
      const step = state.tableData.length + 1;
      const id = Guid.create().toString();
      const newItem: DirectProofsTableItem = {
        step,
        id,
        rawInput: `${firstVariable}, ${secondVariable}`,
        expression,
        friendlyExpression,
        formula,
        comment: { en: 'CR', ru: 'СП' },
      };
      state.selectedIds = [];
      state.tableData = [...state.tableData, newItem];
    },

    eliminateImplication: (state, action: PayloadAction<DirectProofsTableItem[]>) => {
      const items = action.payload;
      if (items.length !== 2) {
        return state;
      }

      const formulas = items.map((item) => item.formula);

      if (!validator.isIEApplicable(formulas[0], formulas[1])) {
        return state;
      }

      const formula = executor.performIE(formulas[0], formulas[1]);
      const expression = converter.convertFormulaToExpression(formula);
      const friendlyExpression = converter.convertFormulaToUserFriendlyExpression(formula);
      const step = state.tableData.length + 1;
      const id = Guid.create().toString();
      const newItem: DirectProofsTableItem = {
        step,
        id,
        expression,
        friendlyExpression,
        formula,
        rawInput: `${items[0].rawInput}, ${items[1].rawInput}`,
        comment: { en: `IE: ${items[0].step}, ${items[1].step}`, ru: `УИ: ${items[0].step}, ${items[1].step}` },
        dependentOn: [items[0].step, items[1].step],
      };

      state.selectedIds = [];
      state.tableData = [...state.tableData, newItem];
    },

    replacePropositionalVariable: (state, action: PayloadAction<{ newVariable: string; oldVariable: string }>) => {
      const { newVariable, oldVariable } = action.payload;
      state.tableData = replacer.replacePropositionalVariableInTableItems(state.tableData, newVariable, oldVariable);
    },
  },
});

export const propositionsDPActions = { ...propositionsDPSlice.actions };
