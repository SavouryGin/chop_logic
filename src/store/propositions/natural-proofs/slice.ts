import converter from 'logic/propositions/converter';
import replacer from 'logic/propositions/replacer';
import { LocalText } from 'types';
import { NPFormulaBase } from 'enums';
import { NaturalProofsTableItem, PropositionsNaturalProofsFlag } from './interfaces';
import { PayloadAction, createSlice } from '@reduxjs/toolkit';
import { propositionsNPInitialState } from './initial-state';

export const propositionsNPSlice = createSlice({
  name: 'Prop-NP',
  initialState: propositionsNPInitialState,
  reducers: {
    setUpFlag: (state, action: PayloadAction<{ flag: PropositionsNaturalProofsFlag; value: boolean }>) => {
      const { flag, value } = action.payload;
      state.flags[flag] = value;
    },

    setSelectedIds: (state, action: PayloadAction<string[]>) => {
      state.selectedIds = action.payload;
    },

    resetError: (state, _action: PayloadAction) => {
      state.error = null;
    },

    setTableData: (state, action: PayloadAction<NaturalProofsTableItem[]>) => {
      state.tableData = action.payload;
    },

    addPremise: (state, action: PayloadAction<string>) => {
      const rawInput = action.payload;
      const expression = converter.convertStringToExpression(rawInput);
      const formula = converter.convertExpressionToFormula(expression);
      const friendlyExpression = converter.convertFormulaToUserFriendlyExpression(formula);
      const step = state.tableData.length + 1;
      const id = crypto.randomUUID();
      const newItem: NaturalProofsTableItem = {
        level: 0,
        id,
        step,
        rawInput,
        expression,
        friendlyExpression,
        formula,
        comment: { en: 'Premise', ru: 'Посылка' },
        formulaBase: NPFormulaBase.Premise,
        assumptionId: null,
        dependentOn: null,
      };
      state.tableData = [...state.tableData, newItem];
    },

    addShortcut: (state, action: PayloadAction<{ rawInput: string; comment: string }>) => {
      const { rawInput, comment } = action.payload;
      const itemsCount = state.tableData.length;
      const expression = converter.convertStringToExpression(rawInput);
      const formula = converter.convertExpressionToFormula(expression);
      const friendlyExpression = converter.convertFormulaToUserFriendlyExpression(formula);
      const step = itemsCount + 1;
      const level = itemsCount > 0 ? state.tableData[itemsCount - 1].level : 0;
      const id = crypto.randomUUID();
      const newItem: NaturalProofsTableItem = {
        level,
        id,
        step,
        rawInput,
        expression,
        friendlyExpression,
        formula,
        comment: { en: comment, ru: comment },
        formulaBase: NPFormulaBase.Shortcut,
        assumptionId: null,
        dependentOn: null,
      };
      state.tableData = [...state.tableData, newItem];
    },

    addAssumption: (state, action: PayloadAction<string>) => {
      const rawInput = action.payload;
      const itemsCount = state.tableData.length;
      const expression = converter.convertStringToExpression(rawInput);
      const formula = converter.convertExpressionToFormula(expression);
      const friendlyExpression = converter.convertFormulaToUserFriendlyExpression(formula);
      const step = itemsCount + 1;
      const id = crypto.randomUUID();
      const level = itemsCount > 0 ? state.tableData[itemsCount - 1].level + 1 : 1;
      const newItem: NaturalProofsTableItem = {
        level,
        id,
        step,
        rawInput,
        expression,
        friendlyExpression,
        formula,
        comment: { en: 'Assumption', ru: 'Гипотеза' },
        formulaBase: NPFormulaBase.Assumption,
        assumptionId: crypto.randomUUID(),
        dependentOn: null,
      };
      state.tableData = [...state.tableData, newItem];
    },

    reiterateStep: (state) => {
      const selectedStep = state.tableData.find((item) => item.id === state.selectedIds[0]);
      if (!selectedStep) {
        return state;
      }
      const index = state.tableData.length + 1;
      const lastItemLevel = state.tableData[state.tableData.length - 1].level;
      const newStep = {
        ...selectedStep,
        id: crypto.randomUUID(),
        step: index,
        comment: { en: `Reiter. ${selectedStep.step}`, ru: `Повтор ${selectedStep.step}` },
        dependentOn: selectedStep.dependentOn?.length ? [...selectedStep.dependentOn, selectedStep.id] : [selectedStep.id],
        level: lastItemLevel,
      };

      state.selectedIds = [];
      state.tableData = [...state.tableData, newStep];
    },

    replacePropositionalVariable: (state, action: PayloadAction<{ newVariable: string; oldVariable: string }>) => {
      const { newVariable, oldVariable } = action.payload;
      state.tableData = replacer.replacePropositionalVariableInNPTableItems(state.tableData, newVariable, oldVariable);
    },

    deleteSteps: (state, _action: PayloadAction<{ isConfirmed: boolean }>) => {
      return state;
    },

    setDependentItems: (state, action: PayloadAction<NaturalProofsTableItem[]>) => {
      state.dependentItems = action.payload;
    },

    setError: (state, action: PayloadAction<LocalText | null>) => {
      state.error = action.payload;
    },

    createDisjunction: (state, _action: PayloadAction<string>) => {
      return state;
    },

    eliminateDisjunction: (state) => {
      return state;
    },

    createConjunction: (state) => {
      return state;
    },

    eliminateConjunction: (state) => {
      return state;
    },

    createNegation: (state) => {
      return state;
    },

    eliminateNegation: (state) => {
      return state;
    },

    createEquivalence: (state) => {
      return state;
    },

    eliminateEquivalence: (state) => {
      return state;
    },

    eliminateImplication: (state) => {
      return state;
    },

    createImplication: (state) => {
      return state;
    },

    exportToXML: (state, _action: PayloadAction<string | undefined>) => {
      return state;
    },

    importFromXML: (state, _action: PayloadAction<{ file: File }>) => {
      return state;
    },
  },
});

export const propositionsNPActions = { ...propositionsNPSlice.actions };
