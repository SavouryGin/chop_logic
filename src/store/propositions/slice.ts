import converter from 'logic/propositions/converter';
import { DirectProofsTableItem, PropositionsFlag, PropositionsInitialState } from './interfaces';
import { PayloadAction, createSlice } from '@reduxjs/toolkit';

export const propositionsInitialState: PropositionsInitialState = {
  flags: {
    isPremiseOpened: false,
    isImplicationCreationOpened: false,
  },
  directProofsTableData: [],
  selectedIds: [],
};

export const propositionsSlice = createSlice({
  name: 'propositions',
  initialState: propositionsInitialState,
  reducers: {
    setUpFlag: (state, action: PayloadAction<{ flag: PropositionsFlag; value: boolean }>) => {
      const { flag, value } = action.payload;
      state.flags[flag] = value;
    },

    setSelectedIds: (state, action: PayloadAction<string[]>) => {
      state.selectedIds = action.payload;
    },

    addPromise: (state, action: PayloadAction<string>) => {
      const rawInput = action.payload;
      const expression = converter.convertStringToExpression(rawInput);
      const formula = converter.convertExpressionToFormula(expression);
      const friendlyExpression = converter.convertFormulaToUserFriendlyExpression(formula);
      const step = state.directProofsTableData.length + 1;
      const id = `proof-step-${step}`;
      const newItem: DirectProofsTableItem = {
        id,
        step,
        rawInput,
        expression,
        friendlyExpression,
        formula,
        comment: { en: 'Premise', ru: 'Посылка' },
      };
      state.directProofsTableData = [...state.directProofsTableData, newItem];
    },

    deleteSteps: (state) => {
      const newSteps = state.directProofsTableData
        .filter((item) => !state.selectedIds.includes(item.id))
        .map((item, index) => {
          return {
            ...item,
            step: index + 1,
            id: `proof-step-${index + 1}`,
          };
        });

      state.selectedIds = [];
      state.directProofsTableData = newSteps;
    },

    reiterateStep: (state) => {
      const selectedStep = state.directProofsTableData.find((item) => item.id === state.selectedIds[0]);
      if (!selectedStep) {
        return state;
      }
      const index = state.directProofsTableData.length + 1;
      const newStep = {
        ...selectedStep,
        id: `proof-step-${index}`,
        step: index,
        comment: { en: `Reiter. ${selectedStep.step}`, ru: `Повтор ${selectedStep.step}` },
      };

      state.selectedIds = [];
      state.directProofsTableData = [...state.directProofsTableData, newStep];
    },

    createImplication: (state, action: PayloadAction<{ firstVariable: string; secondVariable: string }>) => {
      const { firstVariable, secondVariable } = action.payload;
      const expression = converter.convertToICExpression(firstVariable, secondVariable);
      const formula = converter.convertExpressionToFormula(expression);
      const friendlyExpression = converter.convertFormulaToUserFriendlyExpression(formula);
      const step = state.directProofsTableData.length + 1;
      const id = `proof-step-${step}`;
      const newItem: DirectProofsTableItem = {
        step,
        id,
        rawInput: `${firstVariable}, ${secondVariable}`,
        expression,
        friendlyExpression,
        formula,
        comment: { en: 'IC', ru: 'ВИ' },
      };
      state.selectedIds = [];
      state.directProofsTableData = [...state.directProofsTableData, newItem];
    },
  },
});

export const propositionsActions = { ...propositionsSlice.actions };
