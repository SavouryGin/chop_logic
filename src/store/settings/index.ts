import { Language } from 'types';
import { PayloadAction, createSlice } from '@reduxjs/toolkit';
import { SettingsFlag, SettingsInitialState } from './interfaces';

export const settingsInitialState: SettingsInitialState = {
  language: 'en',
  flags: {
    isNavigationOpen: false,
    isSidebarOpen: false,
    isDarkMode: false,
    isFullScreen: false,
    isSettingOpen: false,
    isSoundsEnabled: false,
  },
};

export const settingsSlice = createSlice({
  name: 'settings',
  initialState: settingsInitialState,
  reducers: {
    resetSettings: () => {
      return settingsInitialState;
    },

    toggleFlag: (state, action: PayloadAction<SettingsFlag>) => {
      const currentFlagValue = state.flags[action.payload];
      state.flags[action.payload] = !currentFlagValue;
    },

    setLanguage: (state, action: PayloadAction<Language>) => {
      state.language = action.payload;
    },

    setFullScreenFlag: (state, action: PayloadAction<boolean>) => {
      state.flags.isFullScreen = action.payload;
    },

    setUpFlag: (state, action: PayloadAction<{ flag: SettingsFlag; value: boolean }>) => {
      const { flag, value } = action.payload;
      state.flags[flag] = value;
    },
  },
});

export const settingsActions = { ...settingsSlice.actions };
