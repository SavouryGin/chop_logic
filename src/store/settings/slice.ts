import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { Language, SettingsFlag, SettingsInitialState } from './interfaces';

export const settingsInitialState: SettingsInitialState = {
  language: 'en',
  flags: {
    isNavigationOpen: false,
    isSidebarOpen: false,
    isDarkMode: false,
    isFullScreen: false,
    isSettingOpen: false,
    isSoundsEnabled: true,
  },
};

export const settingsSlice = createSlice({
  name: 'settings',
  initialState: settingsInitialState,
  reducers: {
    resetSettings: () => {
      return settingsInitialState;
    },

    toggleFlag: (state, action: PayloadAction<SettingsFlag>) => {
      const currentFlagValue = state.flags[action.payload];
      state.flags[action.payload] = !currentFlagValue;
    },

    setLanguage: (state, action: PayloadAction<Language>) => {
      state.language = action.payload;
    },
  },
});

export const settingsActions = { ...settingsSlice.actions };
