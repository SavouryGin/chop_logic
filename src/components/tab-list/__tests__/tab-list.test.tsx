import React from 'react';
import renderWithRedux from 'helpers/test-utils/render-with-redux';
import { TabListProps } from 'types';
import { combineReducers } from '@reduxjs/toolkit';
import { fireEvent, screen } from '@testing-library/react';
import { settingsInitialState, settingsSlice } from 'store/settings/slice';
import { testTabs } from '__mocks__/data/tablist';

import TabList from 'components/tab-list';

const testProps: TabListProps = {
  tabs: testTabs,
};

const mockedReducer = combineReducers({
  settings: settingsSlice.reducer,
});

const mockedState = {
  settings: settingsInitialState,
};

describe('Tablist component:', () => {
  it('renders the tablist', () => {
    renderWithRedux(<TabList {...testProps} />, mockedReducer, mockedState);
    expect(screen.getByRole('tablist')).toBeInTheDocument();
  });

  it('renders the correct number of tabs', () => {
    renderWithRedux(<TabList {...testProps} />, mockedReducer, mockedState);
    expect(screen.getAllByRole('tab')).toHaveLength(testTabs.length);
  });

  it('renders only one tabpanel', () => {
    renderWithRedux(<TabList {...testProps} />, mockedReducer, mockedState);
    expect(screen.getByRole('tabpanel')).toBeInTheDocument();
  });

  it('displays titles in the tabs', () => {
    renderWithRedux(<TabList {...testProps} />, mockedReducer, mockedState);
    const tabs = screen.getAllByRole('tab');
    for (let i = 0; i < tabs.length; i++) {
      expect(tabs[i]).toHaveTextContent(testTabs[i].tabTitle.en);
    }
  });

  it('the first tab is selected by default', () => {
    renderWithRedux(<TabList {...testProps} />, mockedReducer, mockedState);
    expect(screen.getByRole('tab', { selected: true })).toHaveAttribute('id', testTabs[0].tabId);
    // All others tabs are not selected
    expect(screen.getAllByRole('tab', { selected: false })).toHaveLength(testTabs.length - 1);
  });

  it('the defaultTabId property can change the selected by default tab', () => {
    const defaultId = testTabs[2].tabId;
    renderWithRedux(<TabList {...testProps} defaultTabId={defaultId} />, mockedReducer, mockedState);
    expect(screen.getByRole('tab', { selected: true })).toHaveAttribute('id', defaultId);
  });

  it('the tabpanel displays only the content for the selected tab', () => {
    renderWithRedux(<TabList {...testProps} />, mockedReducer, mockedState);
    const tabpanel = screen.getByRole('tabpanel');
    expect(tabpanel).toHaveTextContent('Tab content 1');
    expect(tabpanel).not.toHaveTextContent('Tab content 2');
    expect(tabpanel).not.toHaveTextContent('Tab content 3');
    expect(tabpanel).not.toHaveTextContent('Tab content 4');
  });

  it('user can select a new tab by clicking on it', () => {
    renderWithRedux(<TabList {...testProps} />, mockedReducer, mockedState);
    expect(screen.getByRole('tab', { selected: true })).toHaveAttribute('id', testTabs[0].tabId);
    const tabs = screen.getAllByRole('tab');
    fireEvent.click(tabs[1]);
    expect(screen.getByRole('tab', { selected: true })).toHaveAttribute('id', testTabs[1].tabId);
    fireEvent.click(tabs[3]);
    expect(screen.getByRole('tab', { selected: true })).toHaveAttribute('id', testTabs[3].tabId);
  });

  it('the content of the tabpanel changes when user clicks on a new tab', () => {
    renderWithRedux(<TabList {...testProps} />, mockedReducer, mockedState);
    expect(screen.getByRole('tabpanel')).toHaveTextContent('Tab content 1');
    const tabs = screen.getAllByRole('tab');
    fireEvent.click(tabs[1]);
    expect(screen.getByRole('tabpanel')).toHaveTextContent('Tab content 2');
    fireEvent.click(tabs[3]);
    expect(screen.getByRole('tabpanel')).toHaveTextContent('Tab content 4');
  });

  it('the selected tab has different styles', () => {
    renderWithRedux(<TabList {...testProps} />, mockedReducer, mockedState);
    expect(screen.getByRole('tab', { selected: true })).toHaveClass('tab-list__tab', 'tab-list__tab_active');
    const notSelectedTabs = screen.getAllByRole('tab', { selected: false });
    for (const tab of notSelectedTabs) {
      expect(tab).not.toHaveClass('tab-list__tab_active');
      expect(tab).toHaveClass('tab-list__tab');
    }
  });
});
