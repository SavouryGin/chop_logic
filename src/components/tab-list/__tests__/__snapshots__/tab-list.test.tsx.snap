// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tablist component: should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="tab-list"
  >
    <div
      class="tab-list__tabs"
      role="tablist"
    >
      <h3
        aria-selected="true"
        class="tab-list__tab tab-list__tab_active"
        data-testid="tab-1"
        id="tab-1"
        role="tab"
      >
        Title 1
      </h3>
      <h3
        aria-selected="false"
        class="tab-list__tab"
        data-testid="tab-2"
        id="tab-2"
        role="tab"
      >
        Title 2
      </h3>
      <h3
        aria-selected="false"
        class="tab-list__tab"
        data-testid="tab-3"
        id="tab-3"
        role="tab"
      >
        Title 3
      </h3>
      <h3
        aria-selected="false"
        class="tab-list__tab"
        data-testid="tab-4"
        id="tab-4"
        role="tab"
      >
        Title 4
      </h3>
    </div>
    <section
      aria-labelledby="tab-1"
      class="tab-list__content"
      role="tabpanel"
    >
      Tab content 1
    </section>
  </div>
</DocumentFragment>
`;
